Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/tan/SubmissionTanViewModel.kt

```diff
@@ -43,36 +46,52 @@ class SubmissionTanViewModel @AssistedInject constructor(
         currentTan.value = Tan(tan)
     }
 
-    fun onTanSubmit() {
+    fun startTanSubmission() {
         val teletan = currentTan.value
         if (!teletan.isTanValid) {
             Timber.w("Tried to set invalid teletan: %s", teletan)
             return
         }
 
         launch {
-            try {
-                registrationState.postValue(ApiRequestState.STARTED)
-                val request = CoronaTestTAN.PCR(tan = teletan.value)
-                submissionRepository.registerTest(request)
-                registrationState.postValue(ApiRequestState.SUCCESS)
-            } catch (err: CwaWebException) {
-                registrationState.postValue(ApiRequestState.FAILED)
-                registrationError.postValue(err)
-            } catch (err: TransactionException) {
-                if (err.cause is CwaWebException) {
-                    registrationError.postValue(err.cause)
-                } else {
-                    err.report(ExceptionCategory.INTERNAL)
-                }
-                registrationState.postValue(ApiRequestState.FAILED)
-            } catch (err: Exception) {
-                registrationState.postValue(ApiRequestState.FAILED)
+            val pcrTestAlreadyStored = submissionRepository.testForType(CoronaTest.Type.PCR).first()
+            if (pcrTestAlreadyStored != null) {
+                val coronaTestTAN = CoronaTestTAN.PCR(tan = teletan.value)
+                routeToScreen.postValue(
+                    SubmissionNavigationEvents.NavigateToDeletionWarningFragmentFromTan(
+                        consentGiven = false,
+                        coronaTestTan = coronaTestTAN
+                    )
+                )
+            } else {
+                onTanSubmit(teletan)
+            }
+        }
+    }
+
+    private suspend fun onTanSubmit(teletan: Tan) {
+
+        try {
+            registrationState.postValue(ApiRequestState.STARTED)
```

Comment:
lets mark the whole `registrationState` and `registrationError` business for refactoring later

Reply:
Yes I will :)